StateVoicesNational/Spoke
# OSDI action handler and server for import and people, questions, answers, messages

# Fixes 
#597 OSDI record_canvass_helper action handler
#949 API to push contacts into a campaign

## Description

This PR contains an OSDI action handler to submit canvass responses like Question Answers, Tags (aka in VAN terms Survey Question Responses, Activist Codes).  The handler will automatically download the available questions/SQs and tags/ACs and show them in the interaction script editing form as actions.
This can be used to submit data to EveryAction/VAN, Action Network, CiviCRM (in-progress) and other systems that support the OSDI Simple Organizing Profile. ( https://github.com/opensupporter/osdi-docs/pull/337 )

This also contains OSDI Server functionality that lets other OSDI clients read data from Spoke in OSDI format, as well as import contacts with batch mode.
https://github.com/opensupporter/osdi-docs/pull/335

# Checklist:

- [x] I have manually tested my changes on desktop and mobile
- [x] The test suite passes locally with my changes
- [x] If my change is a UI change, I have attached a screenshot to the description section of this pull request
- [ ] [My change is 300 lines of code or less](https://github.com/MoveOnOrg/Spoke/blob/main/CONTRIBUTING.md#your-first-code-contribution), or has a documented reason in the description why it’s longer
- [x] I have made any necessary changes to the documentation
- [ ] My PR is labeled [WIP] if it is in progress

This pull request is more than 300 lines of code.  This is due to the functionality added, as well as incorporating several rounds of feedback, and customer requests.

-------
@joemcl @lperson  I ended up creating a more expansive OSDI implementation.  In addition to supporting people import and collection browsing, I mapped spoke interaction steps, question_response into OSDI Questions and Answers.  I also expose the messages collection, assignments, and users. as well as the relationship links between them.

In addition to automating import, this can be used to simplify integrations moving Q&A data into other systems like VAN et al.  A script in a nightly cron job could suck down new answers, associated people, and do the mapping between the spoke Q&A and the VAN survey questions, or other logic and push them into VAN et al.

I set up a test server with some sample data and a public API for people to play with here:

Browse to http://spoke.dev.joshco.org/osdi

While looking at the HAL browser, try the following walkthrough

1. Click the link for the organization you want to browse
2. See the list of campaigns
3. Choose the campaign you want to browse
4. You should now be at the API Entry Point for the Campaign
5. Click the link for 'osdi:people' and see the first page of people results
6. Choose a person, and look at it's links.  You can navigate to that person's Answers (question responses), or messages
7. Go back to the AEP. (or Click "Go To Entry Point" in the top navbar to start over)
8. Click the link for 'osdi:answers' and see the recently added answers.  From each answer you can navigate to the person (campaign contact) or the question itself (derived from interaction steps)

> You can also start navigating users -> assignments -> messages

More docs here: https://github.com/joshco/Spoke/blob/osdi_key/docs/OSDI.md
